### PR TITLE
sdk/go: Fix TestDependsOnInputs data race

### PR DESCRIPTION
**Pull request by abhinav**
TestDependsOnInputs modified the `monitor` attached to the context
while the context was still using it.
This caused a data race:

```
WARNING: DATA RACE
Read at 0x00c00041d8d8 by goroutine 4280:
  github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*interceptingResourceMonitor).RegisterResource()
      [..]/pulumi/sdk/go/pulumi/resource_test.go:708 +0x44
  github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*Context).registerResource.func1()
      [..]/pulumi/sdk/go/pulumi/context.go:854 +0xd28

Previous write at 0x00c00041d8d8 by goroutine 2880:
  github.com/pulumi/pulumi/sdk/v3/go/pulumi.newInterceptingResourceMonitor()
      [..]/pulumi/sdk/go/pulumi/resource_test.go:671 +0xac
```

Replacing the ResourceMonitorClient appears to be the only way
to get the information that the dependency tracker needs,
but replacing the client after Context creation is a data race.

To address this issue, introduce a new internal RunOption
that allows installing a ResourceMonitorClient middleware
at creation.

    WrapResourceMonitorClient(
        func(ResourceMonitorClient) ResourceMonitorClient,
    ) RunOpton

With this option present, we can decorate the client
with the dependency tracker when the context gets created
rather than afterwards.

This also simplifies the implementation
of the dependency tracker in the process.
That change is on a separate commit
so that it can be reviewed separately.

Refs #10092 
Depends on #11906

### CODE REVIEW

File: sdk/go/pulumi/resource_test.go [hunk 1]
```diff
@@ -548,16 +546,15 @@ func TestDependsOnInputs(t *testing.T) {
...
 			return nil
-		}, WithMocks("project", "stack", &testMonitor{}))
+		}, WithMocks("project", "stack", &testMonitor{}), WrapResourceMonitorClient(depTracker.Wrap))
 		assert.NoError(t, err)
 	})
 
 	t.Run("dynamic", func(t *testing.T) {
 		t.Parallel()
 
+		depTracker := new(dependenciesTracker)
```
Comment (Frassle) [inline sdk/go/pulumi/resource_test.go:556]: nit but nearly everywhere else in the codebase we do `&dependenciesTracker{}`, and it's the same thing but we should probably stay consistent.

**Review by Frassle — APPROVED** [review]

**Comment by abhinav** [comment]
> This PR was included in a batch that successfully built, but then failed to merge into master. It will not be retried.
> 
> Additional information:
> 
> ```json
> Response status code: 422
> {"message":"At least 1 approving review is required by reviewers with write access.","documentation_url":"https://docs.github.com/articles/about-protected-branches"}
> ```

what?

**Comment by abhinav** [comment]
bors retry

**Pull request merged**